Fix the lookahead after [\d or [[:posix] to skip whitespace

NWilson opened the pull request.

zherczeg commented on src/pcre2_compile.c:
```diff
@@ -2498,6 +2498,53 @@ return -1;
+  {
+  c = *ptr;
+
+  if (c == CHAR_BACKSLASH)
```
For me a `switch(*ptr)` looks nicer.

NWilson replied:
OK, done.

carenas commented on testdata/testinput1:
```diff
@@ -6721,4 +6715,16 @@ $/x
+    1
+    -
+\= Expect no match
+    z
```
better check for "space" (`\ \`) as not matching

NWilson replied:
Great, done

NWilson commented:
Actually... there's a bug here (pre-existing), which I should fix.

If you do `[\d-&&[foo]]` in ALT_EXTENDED mode, then the hyphen after the `\d` is not a range.

Either:
* Fix the lookahead, so that it looks for a `]` after the hyphen, OR an operator or `[` in ALT_EXTENDED mode. But that's getting quite a convoluted lookahead
* Get rid of the lookahed; just store a boolean flag that says "we saw a \d...\p or [:posix:] escape" and check that flag when we actually reach the hyphen.

I prefer the second. The lookahead was already buggy (and not matching Perl) so it just seems brittle and unmaintainable, especially as I now realise that I failed to update the lookahead properly for the ALT_EXTENDED case.

NWilson commented:
OK, it's much cleaner now. I've reverted the "improved" lookahead, and removed the lookahead entirely.

Adding a little extra state (on the existing `class_range_state` variable) allows us to track this in a much cleaner way, in my opinion, and we end up matching Perl's behaviour more closely for edge cases like `[\d-\E]`, as was the goal.

Ready for review again.

carenas reviewed (approved):
nice work
